src/sparse/merkle_tree.rs

```diff
@@ -0,0 +1,379 @@
+use crate::common::{AsPathIterator, Buffer, Bytes32, Msb, Node as NodeTrait};
+use fuel_storage::Storage;
+
+use crate::sparse::hash::sum;
+use crate::sparse::{zero_sum, Node, StorageNode};
+
+#[derive(Debug, thiserror::Error)]
+pub enum MerkleTreeError {
+    #[error("Error")]
+    Error(),
+}
+
+pub struct MerkleTree<'storage, StorageError> {
+    root_node: Node,
+    storage: &'storage mut dyn Storage<Bytes32, Buffer, Error = StorageError>,
+}
+
+impl<'a, 'storage, StorageError> MerkleTree<'storage, StorageError>
+where
+    StorageError: std::error::Error + Clone,
+{
+    pub fn new(storage: &'storage mut dyn Storage<Bytes32, Buffer, Error = StorageError>) -> Self {
+        let root_node = Node::create_placeholder();
+        let _ = storage.insert(&root_node.hash(), root_node.as_buffer());
+
+        Self { root_node, storage }
+    }
+
+    pub fn update(&'a mut self, key: &[u8], data: &[u8]) {
+        if data.is_empty() {
+            // If the data is empty, this signifies a delete operation for the given key.
```

Comment:
lacking context here, but why not have a separate `remove(&'a mut self, key: &[u8])` function?

Reply:
lacking context here, but this is non-obvious (potentially dangerous?) behaviour for me. If someone wants to delete a key, why not use `delete` fn?

Reply:
> why not use `delete` fn?

Because this code is at present a port of Celestia's Go implementation.

https://github.com/celestiaorg/smt/blob/6e634fe4424005b4bce55a56a7a559412b4152b3/smt.go#L143-L146

Reply:
@melekes Yes, as John points out, this behaviour comes from the reference implementation in Celestia. I'm incorporating it here so that we have a behavioural match between the two implementations. This specific behaviour is not documented in the Fuel or Celestia SMT spec, so it may be worth making this explicit.

@adlerjohn The `is_empty()` method is idiomatic to Rust when checking if a slice is empty. An `Option` would used when the argument may or may not be present, more analogous to not null or null. Since arrays in Rust are types, we don't actually have null arrays (cxx null pointers), but the array itself may simply have a length of 0. The `update` signature stipulates that the array necessarily be provided (i.e. a slice `&[u8], not an optional slice `Option<&[u8]>`) but we cannot mandate a greater-than-0 length at compile time.